### ISSUE TITLE
Support Drupal's test suite

### ISSUE THREAD

**Issue by legolasbo** [drud/ddev#723]
## Feature request:
While playing around with ddev to try it out as a possible quick-start development environment for Drupal core/module development, I tried running Drupal's PHPUnit test suite. While the actual unit tests work, it's Kernel and Functional tests do not. Having mentored at several Drupal sprints, I've noticed that new (and existing) contributors often struggle with getting the test suite to run locally. It would be great for core/module contributors if ddev could support Drupal's full suite of tests without much configuration. Offering a versatile development environment that can be installed quickly and lowers the barrier to open source contribution.

## How to reproduce this:
Attempt to follow https://www.drupal.org/docs/8/phpunit/running-phpunit-tests and notice that `phantomjs` is not installed on the ddev container. Ignoring that fact and attempting tot get things running for non-javascript tests still seems impossible. Setting the `SIMPLETEST_DB` environment variable in `phpunit.xml.dist` to `mysql://db:db@db/db#testdb` results in `SQLSTATE[HY000] [2002] php_network_getaddresses: getaddrinfo failed: Name or service not known` when attempting to run the tests and configuring it to use an sqlite database fails because `sqlite` is not installed on the container.

## Version:
ddev version:
```
cli   	v0.15.0                        
web   	drud/nginx-php-fpm-local:v1.1.0
db    	drud/mariadb-local:v0.8.0      
dba   	drud/phpmyadmin:v0.2.0         
router	drud/ddev-router:v0.5.0        
commit	v0.15.0                        
domain	ddev.local 
```

docker version:
```
Client:
 Version:	17.12.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	c97c6d6
 Built:	Wed Dec 27 20:03:51 2017
 OS/Arch:	darwin/amd64

Server:
 Engine:
  Version:	17.12.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	c97c6d6
  Built:	Wed Dec 27 20:12:29 2017
  OS/Arch:	linux/amd64
  Experimental:	true
```
config.yml:
```
APIVersion: "1"
name: my-drupal8-site
type: drupal8
docroot: web
php_version: "7.1"
webimage: drud/nginx-php-fpm-local:v1.1.0
dbimage: drud/mariadb-local:v0.8.0
dbaimage: drud/phpmyadmin:v0.2.0
router_http_port: "80"
router_https_port: "443"
provider: default
```

## Related source links or issues:
https://www.drupal.org/docs/8/phpunit

## Complexity
2-3 I think.

**Comment by rickmanelius** [comment]
Hi @legolasbo. We're currently at the Midcamp code sprints and this has come up several times. We fully concur and this is something asked across communities, so we'll review the roadmap and see how quickly we can get to this.

Somewhat related (accessing service ports https://github.com/drud/ddev/issues/691)

**Comment by rfay** [comment]
Some notes on this one:

* phantomjs doesn't actually have to be added to the container to make things work - it could be on the host and mounted in and then linked to somewhere in the path.

Some of the following may help as a workaround, using post-start in config.yaml:

```
hooks:
  post-start:
    - exec: apt-get update
    - exec: apt-get install bzip2 fontconfig
    - exec: wget -qO /tmp/phantomjs.tar.bz2 https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-linux-x86_64.tar.bz2
    - exec: tar -jxf /tmp/phantomjs.tar.bz2 -C /usr/local
    - exec: ln -fs /usr/local/phantomjs-2.1.1-linux-x86_64/bin/phantomjs /usr/local/bin/phantomjs
```

This is just a hacky way to get phantomjs into the container whenever it starts. There's no reason I know of not to provide bzip2 in the container, will need to look at the size of fontconfig; that would remove the first two lines.

Of course the whole phantomjs thing could be done by having them extract the tarball in a subdirectory of the project root (which is mounted). But that's certainly not the only way.

Bottom line: This can be solved a lot of ways, probably a combination of improving docs and adding a bit to the container.

**Comment by rfay** [comment]
One way to work on this is to put the extra config into github.com/drud/quicksprint

**Comment by rfay** [comment]
Working on this because it's important to our sprint support goals. I haven't had an real trouble, and added the required fontconfig package to the web container, so that will go into next release.

**Comment by rfay** [comment]
@legolasbo Could you please try out the script in https://gist.github.com/rfay/830442d7ffcb98d541fc7e9f5b73d616 ? There are a few ways to run it, described there. Based on your feedback we can figure out where it should live. It does not increase the size of our web container, because phantomjs is downloaded as needed, but it's always available for a project. I think it might work. Thanks!

**Comment by legolasbo** [comment]
The script looks promising @rfay, but I'm unable to test it right now. I've written this down on my to-do list, but I've got a very busy week this week so I can't guarantee I'll get to it this week. I'll let you know as soon as I've tried it. 

Initial feedback based solely on having read the script:
`# It assumes a regular Drupal 8 checkout, with the docroot in the project root.` It would be great if the docroot could be configurable, since many people (including myself) use the Drupal composer template to get a quick development environment set up for contributed modules.

**Comment by rickmanelius** [comment]
Adding some flags and getting to our board.

**Comment by rfay** [comment]
IMO this becomes a docs issue. It's more of a HOWTO, the script provided can help most people do what needs to be done. The one thing that (maybe) needs to be done, but could be contributed by a user, is how to debug tests using this technique.

**Comment by rickmanelius** [comment]
Cross-referencing the HOWTO Meta issue #659.

**Comment by rfay** [comment]
The hostname from inside the container is going to be localhost - could you try that? `php ./core/scripts/run-tests.sh --browser --color --url http://localhost --class 'Drupal\Tests\workflows\Functional\WorkflowUiTest'`

**Comment by rfay** [comment]
OH, if you're on v0.16.0, the internal port is port 8080, so would be localhost:8080 - That's going back to port 80 in v0.17.0 (hopefully still this week)

**Comment by scotteuser** [comment]
Beautiful, that works! 

````
./core/scripts/run-tests.sh --browser --color --url http://localhost:8080 --class 'Drupal\Tests\workflows\Functional\WorkflowUiTest'
````

**Comment by rickmanelius** [comment]
Trying to summarize where this landed. It looks like, based on the feedback from @scotteuser that we have a working path forward. However, @rfay mentioned the need to have better documentation coverage. Not sure if this would be best served in Stackoverflow versus our official docs. So two questions.

1. Is this only a documentation issue to get this to a completion?
2. If no, what remains? If yes, any issue with pushing this to Stackoverflow?

I'm betting that this is just a documentation task at this point, and then we can close this out.

**Comment by rfay** [comment]
I think Stack Overflow with an updated github gist would be adequate. There haven't been a lot of calls for this so far.

**Comment by rfay** [comment]
This was declared "good enough" in https://github.com/drud/ddev/issues/723#issuecomment-378670124

So I'm going to close it until further needs arise.

**Comment by scotteuser** [comment]
Just for future reference for anyone, there are a couple more steps which may not be completely obvious. Had to go through this again for the first time in a while and thought I would document my steps a bit better:

```
# Clone the drupal 8 core repo, set the branch to latest
git clone --branch 8.6.x https://git.drupal.org/project/drupal.git drupal8
cd drupal8

# Set up ddev as normal
ddev config
ddev exec composer install

# Visit your Drupal site and run the normal database install

# Enable simpletest module since we don't have sqlite available in the container
ddev exec drush pm-enable simpletest -y

# Get the test setup script from this git issue and run it
# If you have permissions issues you can do `ddev exec chmod SOME-PERMISSION d8test_setup.sh`
ddev exec wget https://gist.githubusercontent.com/rfay/830442d7ffcb98d541fc7e9f5b73d616/raw/b387253e993b75c98805695a15c35c984b46552d/d8test_setup.sh
ddev exec ./d8test_setup.sh

# Upgrade php unit for d8 core php7+
ddev exec composer run-script drupal-phpunit-upgrade

# List all available tests
ddev exec php ./core/scripts/run-tests.sh --php /usr/bin/php --list
# List tests with specific keyword
ddev exec php ./core/scripts/run-tests.sh --php /usr/bin/php --list | grep -lri "Workflow"

# Run a test
ddev exec php ./core/scripts/run-tests.sh --php /usr/bin/php --browser --color --url http://localhost --class 'Drupal\Tests\workflows\Functional\WorkflowUiTest'
```

**Comment by rfay** [comment]
Thanks so much for that! My own preference is to `ddev ssh` and just do everything inside the container; this is fantastic, thanks!

It's easy enough to add sqlite3 either as a one-off or it's possible we should add it to the container, but there are so many reasons not to add *everything* to the container. 

Anyway, thanks so much for your notes!

**Comment by PapaGrande** [comment]
Thanks for this.  Has anyone got it to work with Nightwatch tests?

I'm also having trouble getting composer commands to work in the container when my web root is a nested directory below the code root. Is that possible for running tests?

**Comment by rfay** [comment]
See the excellent ddev-with-phpunit post by @mglaman at https://glamanate.com/blog/running-drupals-phpunit-test-suites-ddev !
